File: BMS-FRONTEND/src/services/busService.js
import api from '../api/axiosConfig';

const searchBuses = async (source, destination, date) => {
  const response = await api.get(`/buses/search?source=${source}&destination=${destination}&travelDate=${date}`);
  return response.data;
};

const getAllBuses = async () => {
  const response = await api.get('/buses');
  return response.data;
};

const getBusById = async (id) => {
  const response = await api.get(`/buses/${id}`);
  return response.data;
};

const addBus = async (busData) => {
  const response = await api.post('/buses', busData);
  return response.data;
};

const deleteBus = async (id) => {
  const response = await api.delete(`/buses/${id}`);
  return response.data;
};

const updateBus = async (id, busData) => {
  const response = await api.put(`/buses/${id}`, busData);
  return response.data;
};

const getBookedSeats = async (busId) => {
  const response = await api.get(`/buses/${busId}/seats`);
  return response.data;
};

const busService = {
  searchBuses,
  getAllBuses,
  getBusById,
  addBus,
  deleteBus,
  updateBus,
  getBookedSeats,
};

export default busService;

File: src/main/java/com/spring/alpha/service/BusService.java
package com.spring.alpha.service;

import com.alpha.busbooking.entity.Bus;
import com.alpha.busbooking.exception.ResourceNotFoundException;
import com.alpha.busbooking.repository.BusRepository;
import com.alpha.busbooking.repository.BookingRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.time.LocalDate;
import java.util.List;

@Service
public class BusService {
    @Autowired
    private BusRepository busRepository;

    @Autowired
    private BookingRepository bookingRepository;

    public List<Bus> searchBuses(String source, String destination, LocalDate travelDate) {
        return busRepository.findBySourceAndDestinationAndTravelDate(source, destination, travelDate);
    }

    public List<Bus> getAllBuses() {
        return busRepository.findAll();
    }

    public Bus getBusById(Long id) {
        return busRepository.findById(id).orElseThrow(() -> new ResourceNotFoundException("Bus not found"));
    }

    public Bus saveBus(Bus bus) {
        return busRepository.save(bus);
    }

    @Transactional
    public Bus updateBus(Long id, Bus updatedBus) {
        Bus existingBus = busRepository.findById(id).orElseThrow(() -> new ResourceNotFoundException("Bus not found"));
        
        existingBus.setBusName(updatedBus.getBusName());
        existingBus.setSource(updatedBus.getSource());
        existingBus.setDestination(updatedBus.getDestination());
        existingBus.setTravelDate(updatedBus.getTravelDate());
        existingBus.setFare(updatedBus.getFare());
        
        // Handle seat updates properly
        int bookedSeats = existingBus.getTotalSeats() - existingBus.getAvailableSeats();
        if (updatedBus.getTotalSeats() < bookedSeats) {
            throw new com.alpha.busbooking.exception.BadRequestException("Total seats cannot be less than already booked seats (" + bookedSeats + ")");
        }
        
        existingBus.setTotalSeats(updatedBus.getTotalSeats());
        existingBus.setAvailableSeats(updatedBus.getTotalSeats() - bookedSeats);
        
        return busRepository.save(existingBus);
    }

    @Transactional
    public void deleteBus(Long id) {
        // Delete all bookings for this bus first to avoid foreign key constraint violations
        bookingRepository.deleteByBusId(id);
        busRepository.deleteById(id);
    }
}
